client/components/tinymce/plugins/media/plugin.jsx

```diff
@@ -54,6 +54,9 @@ function mediaButton( editor ) {
 
 	let nodes = {};
 	let updateMedia; // eslint-disable-line
+	let unsubscribeFromReduxStore;
+	let previousMediaReference;
+	let currentMediaReference;
```

Comment:
I wish these helper variables were defined in a place that's more "local" to the subscription callback.

Reply:
What do you think about extracting at least the subscriber function to hide the media reference variables? I don't really see a way to hide the unsubscribe method from the global `mediaButton` context though.

```diff
diff --git a/client/components/tinymce/plugins/media/plugin.jsx b/client/components/tinymce/plugins/media/plugin.jsx
index a0a1ec96d5..e64a80303d 100644
--- a/client/components/tinymce/plugins/media/plugin.jsx
+++ b/client/components/tinymce/plugins/media/plugin.jsx
@@ -43,6 +43,25 @@ const SIZE_ORDER = [ 'thumbnail', 'medium', 'large', 'full' ];
 let lastDirtyImage = null,
        numOfImagesToUpdate = null;
 
+const createOnMediaChangedListener = ( reduxStore, callback ) => {
+       let previousMediaReference, currentMediaReference;
+
+       return reduxStore.subscribe( () => {
+               const siteId = getSelectedSiteId( reduxStore.getState() );
+
+               if ( ! siteId ) {
+                       return;
+               }
+
+               previousMediaReference = currentMediaReference;
+               currentMediaReference = getMedia( reduxStore.getState(), siteId );
+
+               if ( previousMediaReference !== currentMediaReference ) {
+                       callback();
+               }
+       } );
+};
+
 function mediaButton( editor ) {
        const store = editor.getParam( 'redux_store' );
 
@@ -54,9 +73,7 @@ function mediaButton( editor ) {
 
        let nodes = {};
        let updateMedia; // eslint-disable-line
-       let unsubscribeFromReduxStore;
-       let previousMediaReference;
-       let currentMediaReference;
+       let unsubscribeFromMediaChanges;
 
        const getSelectedSiteFromState = () => getSelectedSite( getState() );
 
@@ -881,19 +898,8 @@ function mediaButton( editor ) {
        editor.on( 'touchstart touchmove touchend', selectImageOnTap() );
 
        editor.on( 'init', function () {
-               unsubscribeFromReduxStore = store.subscribe( () => {
-                       const siteId = getSelectedSiteId( store.getState() );
-                       if ( ! siteId ) {
-                               return;
-                       }
+               unsubscribeFromMediaChanges = createOnMediaChangedListener( store, updateMedia );
 
-                       previousMediaReference = currentMediaReference;
-                       currentMediaReference = getMedia( store.getState(), siteId );
-
-                       if ( previousMediaReference !== currentMediaReference ) {
-                               updateMedia();
-                       }
-               } );
                editor.getDoc().addEventListener( 'load', resizeOnImageLoad, true );
                editor.dom.bind( editor.getDoc(), 'dragstart dragend', hideDropZoneOnDrag );
                editor.selection.selectorChanged( '.wp-caption', removeEmptyCaptions );
@@ -912,7 +918,7 @@ function mediaButton( editor ) {
                } );
                nodes = {};
 
-               unsubscribeFromReduxStore();
+               unsubscribeFromMediaChanges();
                editor.getDoc().removeEventListener( 'load', resizeOnImageLoad );
        } );
```

Reply:
The `unsubscribe` variable is OK, I was referring to the `previous` and `current` ones. Apologies for not being clear enough.

In your example, I would change the `createOnMediaChangedListener` function to only return a listener function: the `store.subscribe` call would happen outside it. Then the `previous` and `current` refs are local to the the `create` function.

When manipulating streams of events, it helps to think how one would implement it in RxJS. The store would emit a stream of events. For every event, we want to get the current `getMedia` result (`map` operator) and then coalesce the stream: emit a value only when the source value changes (don't know what operator that is). Their abstractions and generic helpers and operators are great and will help you structure your code better even if it doesn't use the library.

Reply:
Pushed changes in cdebb4b, let me know if that'd work

Reply:
That's great, thank you! 🥇